// appium-xcuitest-driver/build/lib/wda/iproxy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.iProxy = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("../logger"));

var _utils = require("./utils");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _appiumSupport = require("appium-support");

var _teen_process = require("teen_process");

const IPROXY_TIMEOUT = 5000;

const iproxyLog = _appiumSupport.logger.getLogger('iProxy');

class iProxy {
  constructor(udid, localport, deviceport) {
    _logger.default.debug(`Starting iproxy to forward traffic from local port ${localport} to device port ${deviceport} over USB ` + `for the device ${udid}`);

    this.expectIProxyErrors = true;
    this.iproxy = new _teen_process.SubProcess('iproxy', [localport, deviceport, udid], {
      detached: true,
      stdio: ['ignore', 'pipe', 'pipe']
    });
  }

  async start() {
    this.expectIProxyErrors = true;
    return await new _bluebird.default((resolve, reject) => {
      this.iproxy.on('exit', code => {
        _logger.default.debug(`iproxy exited with code '${code}'`);

        if (code) {
          return reject(new Error(`iproxy exited with code '${code}'`));
        }
      });
      this.iproxy.on('output', (stdout, stderr) => {
        if (this.expectIProxyErrors) {
          return;
        }

        let out = stdout || stderr;

        for (let line of out.split('\n')) {
          if (!line.length) {
            continue;
          }

          if (line.includes('Resource temporarily unavailable')) {
            _logger.default.debug('Connection to WDA timed out');
          } else {
            iproxyLog.debug(line);
          }
        }
      });
      return (async () => {
        try {
          await this.iproxy.start(IPROXY_TIMEOUT, true);
          resolve();
        } catch (err) {
          _logger.default.error(`Error starting iproxy: '${err.message}'`);

          reject(new Error('Unable to start iproxy. Is it installed?'));
        }
      })();
    });
  }

  async quit() {
    await (0, _utils.killProcess)('iproxy', this.iproxy);
    this.expectIProxyErrors = true;
  }

}

exports.iProxy = iProxy;
var _default = iProxy;
exports.default = _default;require('source-map-support').install();
