// appium-ios-simulator/build/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BOOT_COMPLETED_EVENT = exports.hasSSLCert = exports.uninstallSSLCert = exports.installSSLCert = exports.simExists = exports.endAllSimulatorDaemons = exports.killAllSimulators = exports.getDeviceString = exports.getSimulator = void 0;

var sim = _interopRequireWildcard(require("./lib/simulator"));

var utils = _interopRequireWildcard(require("./lib/utils"));

var sim6 = _interopRequireWildcard(require("./lib/simulator-xcode-6"));

const getSimulator = sim.getSimulator,
      getDeviceString = sim.getDeviceString;
exports.getDeviceString = getDeviceString;
exports.getSimulator = getSimulator;
const killAllSimulators = utils.killAllSimulators,
      endAllSimulatorDaemons = utils.endAllSimulatorDaemons,
      simExists = utils.simExists,
      installSSLCert = utils.installSSLCert,
      uninstallSSLCert = utils.uninstallSSLCert,
      hasSSLCert = utils.hasSSLCert;
exports.hasSSLCert = hasSSLCert;
exports.uninstallSSLCert = uninstallSSLCert;
exports.installSSLCert = installSSLCert;
exports.simExists = simExists;
exports.endAllSimulatorDaemons = endAllSimulatorDaemons;
exports.killAllSimulators = killAllSimulators;
const BOOT_COMPLETED_EVENT = sim6.BOOT_COMPLETED_EVENT;
exports.BOOT_COMPLETED_EVENT = BOOT_COMPLETED_EVENT;require('source-map-support').install();
